apps/emqx_rule_engine/src/emqx_rule_engine.erl

```diff
@@ -542,3 +548,21 @@ get_egress_bridges(Actions) ->
         emqx_bridge_resource:bridge_id(BridgeType, BridgeName)
      || {bridge, BridgeType, BridgeName, _ResId} <- Actions
     ].
+
+%% For allowing an external application to add extra "built-in" functions to the
+%% rule engine SQL like language. The module set by
+%% set_extra_functions_module/1 should export a function called
+%% handle_rule_function with two parameters (the first being an atom for the
+%% the function name and the second a list of arguments). The function should
+%% should return the result or {error, no_match_for_function} if it cannot
+%% handle the function. See '$handle_undefined_function' in the emqx_rule_funcs
+%% module. See also callback function declaration in emqx_rule_funcs.erl.
+
+-spec extra_functions_module() -> module() | undefined.
+extra_functions_module() ->
+    persistent_term:get({?MODULE, extra_functions}, undefined).
```

Comment:
> Minor nit: I wonder if we should make this handle a list of modules, so we may extend the rule engine using more than one module?

I was thinking about this as well. However, I think since this is not a user facing function and it is easy to change so that we store a list when that is needed in the future, we can leave it as it is right now. Do you think that sounds reasonable? If we get many function modules, we probably want to solve this some other way since it becomes inefficient to traverse a long list (perhaps a map (function name -> function)).